include/kodlab_mjbots_sdk/attitude.h

```diff
@@ -0,0 +1,334 @@
+ * 
+ */
+
+#pragma once
+
+#include <Eigen/Geometry>
+#include "kodlab_mjbots_sdk/pi3hat.h"
+#include "kodlab_mjbots_sdk/common_header.h"
+#include "kodlab_mjbots_sdk/rotations.h"
+#include <iostream>
+#include <iomanip>
+
+namespace kodlab
+{
+
+/**
+ * @brief Attitude object storing inertial data.
+ * @tparam Scalar[optional] numeric type
+ */
+template<typename Scalar = float>
+class Attitude
+{
+
+private:
+  /**
+   * @brief Inputted attitude quaternion before world offset rotation
+   */
+  Eigen::Quaternion<Scalar> att_quat_raw_ = {1, 0, 0, 0};
+
+  /**
+   * @brief Attitude quaternion in world frame
+   */
+  Eigen::Quaternion<Scalar> att_quat_ = {1, 0, 0, 0};
+
+  /**
+   * @brief Attitude euler angles in world frame
+   * @note Always follow the default <b>Extrinsic X-Y-Z> convention
+   */
+  mutable kodlab::ValidatedCache<rotations::EulerAngles<Scalar>> att_euler_;
+
+  /**
+   * @brief Attitude rotation matrix in world frame
+   */
+  mutable kodlab::ValidatedCache<Eigen::Matrix3<Scalar>> att_rot_mat_;
+
+  /**
+   * @brief Angular velocity about <code>[x, y, z]</code> axes (default deg/s)
+   */
+  Eigen::Vector3<Scalar> ang_rate_ = {0, 0, 0};
+
+  /**
+   * @brief Linear acceleration along <code>[x, y, z]</code> axes (default
+   *        m/s^2)
+   */
+  Eigen::Vector3<Scalar> accel_ = {0, 0, 0};
+
+  /**
+   * @brief Gyroscope bias about <code>[x, y, z]</code> axes (default deg/s)
+   */
+  Eigen::Vector3<Scalar> ang_rate_bias_ = {0, 0, 0};
+
+  /**
+   * @brief Attitude uncertainty about the raw attitude
+   */
+  Eigen::Quaternion<Scalar> att_uncertainty_ = {1, 0, 0, 0};
+
+  /**
+   * @brief Gyroscope bias uncertainty about <code>[x, y, z]</code> axes
+   * (default deg/s)
+   */
+  Eigen::Vector3<Scalar> ang_bias_uncertainty_ = {0, 0, 0};
+
+  /**
+   * @brief World frame rotation offset
+   * @note Useful to change the direction of gravity
+   */
+  Eigen::Quaternion<Scalar> world_offset_ = {1, 0, 0, 0};
+
+  /**
+   * @brief Invalidates cached data
+   */
+  void InvalidateCached()
+  {
+    att_euler_.invalidate();
+    att_rot_mat_.invalidate();
+  }
+
+public:
+  /**
+   * @brief Construct a new \c Attitude object.
+   */
+  Attitude() = default;
+
+  /**
+   * @brief Construct a new \c Attitude object with populated data.
+   *
+   * @param att_quat_in attitude quaternion
+   * @param ang_rate_in angular velocity about <code>[x, y, z]</code> axes
+   * @param accel_in linear acceleration along <code>[x, y, z]</code> axes
+   * @param world_offset_in world frame rotation offset (default identity)
+   */
+  Attitude(const Eigen::Quaternion<Scalar> &att_quat_in,
+           const Eigen::Vector3<Scalar> &ang_rate_in,
+           const Eigen::Vector3<Scalar> &accel_in,
+           const Eigen::Quaternion<Scalar> &world_offset_in = {1, 0, 0, 0})
+      : att_quat_raw_(att_quat_in),
+        att_quat_(world_offset_in * att_quat_in),
+        ang_rate_(ang_rate_in),
```

Comment:
If we have this constructor and the update function to match, we could have "invalid" data in the uncertainty and bias members. Likely we should make each an ValidatedCache or make a "estimator" or "error" struct and make that an invalid cache. Alternatively we could make a simpler class and derive this type of update. I think the intention was to allow the user to use simpler versions in the case we were using a different IMU?

Reply:
Yes the intention is for the object to support simpler implementations as well.  I think a note in the constructor comment warning of this intention should suffice.

Reply:
A comment is good starting point, might add issue for dealing with it more carefully in the future

Reply:
Support for alternative Euler conventions provided in bdfce1322a036db5ad1a8e1a939c31b6042e2025